---
title: "PathAnalyser"
author:
  - name: Rishab Kaushik
  - name: Taniya Pal
  - name: Ozlem Karadeniz
  - name: Yi-Hsuan Lee
  - name: Anisha Thind
  - name: Justice Iheanyichi
package: PathAnalyser
abstract: >
  PathAnalyser provides functionality for assessing ER and HER2 pathway activity
  in breast cancer transcriptomic datasets. The package enables classification 
  of samples in transciptomic datasets according to pathway activity by using a 
  gene expression signature associated with a pathway, a list of genes, which 
  vary in expression depending on the pathway activity. These transcriptional 
  signatures have been shown to have clinical predictive value, for example, ER 
  and HER2 gene signatures have been associated with molecular sub-types, 
  prognosis and treatment reponse in breast cancer. Although this package was 
  original developed to assess ER and HER2 pathway activity in breast cancer 
  transcriptomic datasets, the package functionality could be applied to 
  transcriptomic datasets and gene signatures outside the context of breast 
  cancer. In this vignette, we will describe how to use the PathAnalyser package 
  with microarray and RNA-seq expression transcriptomic datasets and gene 
  expression signatures associated with a specific pathway activity.
output: 
  BiocStyle::html_document:
    toc: TRUE
    toc_float: TRUE
vignette: >
  %\VignetteIndexEntry{PathAnalyser}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)
```
# Installation and setup
To install a local source version of the package, type the following in R:
```{r eval=FALSE}
install.packages("~/PathAnalyser", repo=NULL, type="source")
```

Alternatively, the package can be installed directly from GitHub using the
devtools package in R:
```{r eval=FALSE}
install.packages("devtools")
devtools::install_github("a-thind/PathAnalyser")
```

Once installation is completed, load the library to start using PathAnalyser:
```{r setup}
library(PathAnalyser)
```

# Data formats
The classification algorithm of PathAnalyser requires two types of input data:

1. **Gene expression matrix** - a gene-by-sample gene expression matrix, with 
row names corresponding to gene names and column names corresponding to sample 
names / IDs.
2. **gene signature data frame** - a data frame of expression values corresponding 
to a list of genes associated with a pathway. The first column contains gene 
names and the second column contains expression values: -1 for down-regulated or
1 for up-regulated genes when a specific pathway is active.

## Input files
The two data types described above (gene expression matrix and gene signature 
dataframe) can be created from two types of input files:

1. *Gene expression matrix file* - a tab-delimited text file containing a gene 
expression matrix, where rows represent genes and columns samples of a gene 
expression data set. The gene expression matrix text file can contain expression 
values from RNAseq or normalized microarray data. The image below is a snapshot 
of an example expression dataset file:

![Subset of an gene expression matrix text file](expr_dataset_example.png)

The first line contains the sample names and the first column contains gene 
names. All other tab-separated fields represent expression values for each gene
for each sample from eith RNASeq or microarray experiments. 

**Note: PathAnalyser does not provide functionality for unnormalized microarray 
data, so the user must ensure the microarray data is normalized prior to 
performing classification by PathAnalyser.**  

2. *Gene signature files* - this data is provided as two text files containing 
a list of genes provided in the [gene set file format](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GRP:_Gene_set_file_format_.28.2A.grp.29) (using .grp file extension) - one for the up-regulated gene set 
and another containing the down-regulated gene set of the pathway signature. 
A plethora of pathway gene signatures in the gene set file format can be found at 
[MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp). The snapshot below
shows the first 30 lines of the gene set file for the up-regulated gene set of 
the ERBB2 (HER2) gene signature.  

![Gene set file format containing either the up- or down-regulated gene sets of the gene signature](up_gene_signature.png)

**Note: `read_signature_data` from PathAnalyser only accepts gene signatures 
that have the up-regulated and down-regulated gene sets in separate gene set 
format text files (with file extension .grp). The user is expected to provide one 
up-regulated and one down-regulated gene set file for a pathway gene signature.**


## Built-in datasets and gene 
### Gene expression signatures
PathAnalyser contains two built-in gene signatures for active ER and HER2 
pathways that can be used to assess ER and HER2 pathway activity in a breast 
cancer transcriptomic data set. These signatures are accessible by typing the 
following in R:

# Quick Start
Here are the most basic step of the standard workflow for assessing ER / HER2 
pathway activity in a transcriptomic data set using PathAnalyser. Each of these 
steps will be discussed in more detail in sections below.
```{r eval=FALSE}

# Load transcriptomic dataset (gene expression matrix of samples)

data_se <- read_expression_data("/Users/taniyapal/Documents/Group Project/TCGA_unannotated.txt")

# read signature data
sig_df <- read_signature_data("/Users/taniyapal/Documents/Group Project/ESR1_UP.v1._UP.csv","/Users/taniyapal/Documents/Group Project/ESR1_DN.v1_DN.csv" )



transform_matrix(data_se)


## QC and pre-process data

# Transforming matrix with log cpm transformation and sanity check of the transformation
normalized_se <- log_cpm_transformation(data_se)

up_genes_df <- sig_df[sig_df$expression == 1, ]
down_genes_df <- sig_df[sig_df$expression == -1, ]
data_se <- check_signature_vs_dataset(normalized_se, up_genes_df, down_genes_df)



# classify samples using quartile thresholds by default
classes_df <- classify_GSVA_percent(sig_df, data_se)

# assess accuracy of classification
true_labels_df <- read.table("Sample_labels.txt", header=TRUE,sep = "\t")
confusion_matrix <- calculate_accuracy(true_labels_df, classes_df)
```


# Overview of PathAnalyser functionality


The PathAnalyser package functionality can be sub-divided into 5 broad 
categories corresponding to a typical workflow for assessing pathway activity of 
samples:

1.  Input functions for expression data and gene signatures
2.  Quality control and pre-processing of input data
3.  Classification of samples by pathway activity
4.  Evaluation of classification
5.  Visualization

The features unique to PathAnalyser package are:

1) It takes into account multiple sample for the analysis. This is an addition to GSVA package features, where a single sample is
taken into account for analysis.

2) It classifies the IDs/symbols into active and inactive pathway. This classification is based on the two matrices representating up and down regulated expression of genes respectively. The matrices consists of GSVA scores derived after GSVA analysis. It also returns a prediction accuracy score and ROC curve for verification of the accuray of the prediction.

3) It returns a PCA plot for visualization of the categories (Active/Inactive) the gene symbols/IDs have been classified into.


# Built-in datasets and gene 
## Gene expression signatures
PathAnalyser contains two gene signatures for active ER and HER2 pathways that 
can be used to assess ER and HER2 pathway activity in a breast cancer 
transcriptomic data set. These signatures are accessible by typing the following 
in R:

```{r}
data("ER_sig_df")
data("HER2_sig_df")
```
### ER signature
The built-in ER signature `ER_sig_df` is a data frame containing the names of 
160 differentially expressed genes which constitute the transcriptional 
signature of ER pathway activation. This signature was obtained from the 
sensitivity to endocrine therapy genome index defined by Symanns et al. (2011). 
The total number of genes that are up-regulated (denoted with an expression 
value of 1) and down-regulated (denoted with an expression value of -1) in this 
gene signature are 101 and 59 genes respectively.
```{r}
dim(ER_sig_df)
head(ER_sig_df)
# up-regulated genes are given an expression value of 1
ER_up_sig <- ER_sig_df[ER_sig_df$expression == 1 ,]
dim(ER_up_sig)
# down-regulated genes are given an expression value of -1
ER_dn_sig <- ER_sig_df[ER_sig_df$expression == -1 ,]
dim(ER_dn_sig)
# for more details
?ER_sig_df
```

### HER2 signature
A gene expression signature for HER2 (ERBB2) pathway activation `HER2_sig_df` is 
also provided by PathAnalyser. This gene signature was obtained from Molecular 
Signatures Database ([MSigDB](https://www.gsea-msigdb.org/gsea/msigdb/index.jsp)) 
by combining the up-regulated (gene set:
[ERBB2_UP.V1_UP](https://www.gsea-msigdb.org/gsea/msigdb/cards/ERBB2_UP.V1_UP.html)
) and down-regulated (gene set:
[ERBB2_UP.V1_DN](https://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?geneSetName=ERBB2_UP.V1_DN)
) gene sets available at MSigDB, originating from transcriptomic profiling of a 
MCF-7 breast cancer cell line, positive for the estrogen receptor (ER) but 
genetically engineered to express the ERBB2 (HER2) receptor. 

The names of 387 differentially expressed genes are in the HER2 signature data 
frame provided by PathAnalyser, of which 190 are up-regulated 
(denoted by an expression of 1) and 197 are down-regulated (denoted by an 
expression of -1).
```{r}
dim(HER2_sig_df)
head(HER2_sig_df)
# up-regulated genes are given an expression value of 1
HER2_up_sig <- HER2_sig_df[HER2_sig_df$expression == 1 ,]
dim(HER2_up_sig)
# Down-regulated genes are given an expression value of -1
HER2_dn_sig <- HER2_sig_df[HER2_sig_df$expression == -1 ,]
dim(HER2_dn_sig)
# for more details
?HER2_sig_df
```
## Gene expression datasets
PathAnalyser also contains two built-in gene expression matrices each containing
RNA-seq raw read counts for primary breast cancer samples obtained from 20 
individuals (cases). Data for these matrices were obtained from The Cancer 
Genome Atlas ([TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga)).
Each expression matrix contains 20,124 genes.
```{r}
data("ER_data_se1")
data("HER2_data_se1")
# column names represent case (sample) IDs from TCGA
```

### ER dataset 1
The ER data set `ER_data_se1` contains RNASeq raw read counts for 20 primary 
breast cancer samples, 10 of which have ER pathway activity (ER+) and 10 which 
have inactive ER pathway activity (ER-).
```{r}
dim(ER_data_se1)
# Expression data for first 6 genes
head(ER_data_se1)
```

### HER2 dataset 1
Similarly, the HER2 data set `HER2_data_se1` contains RNASeq raw read counts for 
20 primary breast cancer samples, 10 of which have HER2 (ERBB2) pathway activity 
(HER2+) and 10 which have inactive HER2 pathway activity (HER2-).

```{r}
dim(HER2_data_se1)
# Expression data for first 6 genes
head(HER2_data_se1)
```


# Reading input data using PathAnalyser
There are two types of input required for pathway activity assessment using 
PathAnalyser:

1.  gene expression data matrices
2.  gene expression signatures

## Gene expression data
The `read_input_file` function reads gene expression data from  
a file with an delimiter, the function checks the delimiter of the
file and reads the file accordingly. It returns an ouput with gene symbols/IDs as the rownames and columns representating each sample 
IDs in the dataset.
```{r eval=F}
data_se <- read_expression_data("/Users/taniyapal/Documents/Group Project/TCGA_unannotated.txt")
data_se<-data_se[,1:200]
print(data_se[1:5, 1:5])


```

## Gene signature data
Gene signature files can be read by the `read_signature_data` function. It returns a dataframe with signature IDs/symbols in the
first column and their expression pattern is represented as up r
regulated or down regulated by +1 and -1 respectively.
```{r eval=F}
sig_df <- read_signature_data("/Users/taniyapal/Documents/Group Project/ESR1_UP.v1._UP.csv", "/Users/taniyapal/Documents/Group Project/ESR1_DN.v1_DN.csv")
head(sig_df)
```



## Quality control and pre-processing of data
The log_cpm_transformation function transforms the raw data by the method of log cpm
and returns the transformed data. It also performs sanity check of the 
transformation by returning box plots for visualization of distribution of data before
and after transformation.
```{r}
# using toy data set as expression matrix
data("ER_data_se1")
data_se <- ER_data_se1 
normalized_se <- log_cpm_transformation(data_se)

data_se <- check_signature_vs_dataset(normalized_se, ER_sig_df)
```

# Classifying samples by pathway activity using gene signatures
## Overview of pathway activity classification
PathAnalyser assess the pathway activity of each sample in a gene expression 
data set using the Gene set variation analysis (GSVA) method 
([Hänzelmann et al. 2013][References]) from the 
[GSVA](https://www.bioconductor.org/packages/release/bioc/html/GSVA.html) 
bioconductor library to estimate the abundance of expression of the up-regulated 
and down-regulated gene sets of a given pathway's gene signature by calculating 
GSVA scores. The estimated expression abundance for both these gene sets is used 
to check expression consistency with the gene signature and infer pathway
activity for a given sample in the transcriptomic data set.


The classification algorithm implemented in PathAnalyser uses four thresholds 
for the classification algorithm (two for each gene set of the pathway 
signature):

1. **High abundant expression for up-regulated gene set**
2. **Low abundant expression for up-regulated gene set**
3. **High abundant expression for down-regulated gene set**
4. **Low abundant expression for down-regulated gene set**

Samples are ordered by expression abundance of up-regulated and down-regulated 
genes of the gene signature separately using the GSVA scores. Those samples 
which have the most abundant expression of the up-regulated gene set 
(exceeding threshold 1) show evidence of activation of the pathway but are only 
classified as having the pathway "*Active*", if the same samples show a low 
abundance of expression of the down-regulated genes (below threshold 4) of the 
gene signature i.e. demonstrating consistency between evidence from up- and 
down-regulated gene sets of the gene signature. Similarly, samples with evidence 
of pathway inactivation from the up-regulated gene set of the gene signature 
i.e. the least abundant expression of the up-regulated genes of the gene 
signature (below threshold 2) and which also show evidence of pathway 
inactivation from the down-regulated gene set of the signature i.e. 
most abundance expression of the down-regulated gene set of the gene signature
(exceeding threshold 3), demonstrate evidence of consistency of pathway 
inactivation from both gene sets and therefore are classified as "*Inactive*".
All other samples are classified as "*Uncertain*" since there is insufficient 
evidence for pathway activation or inactivation.

In PathAnalyser, the GSVA scores thresholds for classifying a sample as "Active" 
(consistent with the gene signature) and "Inactive" (inconsistent with the gene 
signature) can be set manually and depend on the user's pathway activity 
classification requirements. 

PathAnalyser provides two functions for classifying samples using an absolute 
GSVA scores thresholds `classify_GSVA_abs` or percentile thresholds which is 
effectively rank-based `classify_GSVA_percent` using percentiles (default at 25%)
, as well as a visualization function to view the distributions of the GSVA 
scores for both gene sets `gsva_scores_dist`, which will be discussed below.

## Visualising GSVA Score Distribution
PathAnalyser provides `gsva_scores_dist` method to visualize the GSVA score 
distribution for the abundance of expression of the up-regulated and 
down-regulated gene sets of the gene signature for each sample. The resulting 
density plot usually follows a bimodal distribution of GSVA scores for each 
sample for the up-regulated and down-regulated gene sets, reflective of the 
binary nature of pathway activity and gene expression broadly (samples will 
tend to have either very high or low expression of the up-regulated and 
down-regulated gene set resulting in two modes or peaks for each gene set). 
Figure \@ref(fig:plot) shows two peaks for each gene set when running 
`gsva_scores_dist` with logCPM-normalized `ER_data_se1` gene expression matrix. For 
both gene sets, the peak with the highest GSVA scores represents samples with 
*high* abundance of expression of the gene set and the lower score peak 
represents samples with *low* abundance of expression of the gene set.
```{r plot, fig.cap="Density plots showing the distribution of GSVA scores for samples in logCPM-normalised `ER_data_se1` data set for the up-regulated (left) and down-regulated gene set (right) of the pathway signature using `gsva_scores_dist` function."}
gsva_scores_dist(ER_sig_df, normalized_se)
```

### Determining thresolds for classification
As the two peaks represent represent low expression abundance and high 
abundance expression, the distribution of GSVA scores that constitute the "valley"
i.e. the area between the two peaks (fig. \@ref(fig:thresholds)), represent 
samples that exhibit relatively weaker evidence of differential expression 
abundance relative to other samples and the algorithm would classify these 
samples as having "Uncertain" pathway activity. Because the peaks represent the 
mode GSVA scores (one for high and low expression abundance), thresholds should 
be selected between these two peak to enable the algorithm to classify samples 
around these modes as being consistent or inconsistent with the pathway 
signature.

As mentioned above, there are four thresholds that can be tuned by the user:
the high and low expression abundance for the up-regulated and down-regulated
gene sets. For example, a user could set the high expression abundance thresholds 
for the up-regulated gene set to -0.2 for the low expression threshold for both gene 
sets, and 0.2 for the high expression thresholds for both gene sets. Samples 
with GSVA scores between the these thresholds for both the up-regulated and down
-regulated gene-set would be considered by the classification algorithm as having 
"Uncertain" pathway activity.

```{r thresholds, fig.cap="GSVA scores distributions of the samples for the down-regulated (left) and up-regulated gene set (right) of the pathway gene signature, showing absolute thresholds of -0.2 and 0.2 for characterising low and high expression abundance of both gene sets respectively. The expression data set used is the logCPM-normalised `ER_data_se1` gene expression data set."}
plot <- gsva_scores_dist(ER_sig_df, data_se)
# Add thresholds on plot
library(ggplot2)
data_threshs <- data.frame(Geneset=c("Up", "Down"), vline=c(-0.2, 0.2))
plot + geom_vline(xintercept=data_threshs$vline, linetype=2)
```

Shrinking the distance between the high and low expression thresholds would 
result in fewer "Uncertain" pathway activity classifications 
(fig. \@ref(fig:relaxed)), because more samples would meet the consistency/ 
inconsistency expression thresholds for the pathway signature (high expression 
thresholds would be lower, low expression thresholds would be higher for each 
gene set).

```{r relaxed, fig.cap="GSVA scores distributions of samples with more relaxed thresholds for assessing expression consistency of the down-regulated gene set (left) and the up-regulated gene set (right) with the pathway gene expression signature. Low and high expression abundance thresholds for the down-regulated gene set (left) are set to -0.1 and 0.1 respectively, while low and high expression abundance thresholds for the up-regulated gene set (right) are also set to -0.1 and 0.1 respectively. Data is generated from running GSVA on logCPM-normalised `ER_data_se1` expression dataset with ER signature (`ER_sig_df`)."}
# more relaxed thresholds, fewer uncertain labels
data_threshs <- data.frame(Geneset=c("Up", "Down"), 
                           vline=c(-0.1, 0.1))
plot + geom_vline(xintercept=data_threshs$vline, linetype=2)
```

Conversely, expanding the distance between the thresholds would increase the 
frequency of "Uncertain" pathway activity classifications, as the number of samples
meeting the thresholds for consistency and inconsistency would be reduced.

```{r stringent, fig.cap="Density plots for GSVA scores distributions for samples with relatively more stringent thresholds for assessing consistency of the up-regulated gene set (left) and down-rgulated gene set (right) with the pathway signature. The low and high expression abundance thresholds are -0.4 and 0.4 for both up-regulated and down-regulated gene sets respectively. Data is generated from running GSVA on logCPM-normalised `ER_data_se1` expression dataset with ER signature (`ER_sig_df`)."}
# more stringent thresholds, greater uncertain labels
data_threshs <- data.frame(Geneset=c("Up", "Down"), vline=c(-0.4, 0.4))
plot + geom_vline(xintercept=data_threshs$vline, linetype=2)
```

## Classification using absolute GSVA score thresholds
As the distribution of GSVA scores tends to be biomodal rather than Guassian, 
the user may prefer to use absolute GSVA thresholds for checking consistency of 
expression abundance of each gene set with pathway signature for each sample.
PathAnalyser provides the `classify_GSVA_abs` function for classifying samples 
by pathway activity using absolute GSVA score thresholds. The thresholds for
high expression abundance and low expression abundance of the up-regulated 
gene-set of the pathway signature are specified with `up_thresh.high` and 
`up_thresh.low` parameters. Similarly thresholds for high expression abundance
and low expression abundance for the down-regulated gene set of the signature 
are provided by the `dn_thresh.high` and `dn_thresh.low` parameters respectively.
Note that absolute thresholds are required from the user when running 
`classify_GSVA_abs` and can be positive or negative numbers. A summary of the 
classification is printed to the console highlighting the number of samples in 
each pathway activity class ("Active", "Inactive" and "Uncertain") and the 
number of samples in total.
```{r}
classes_df <- classify_GSVA_abs(ER_sig_df, data_se, 
                                    up_thresh.low=-0.2, up_thresh.high=0.2, 
                                    dn_thresh.low=-0.2, dn_thresh.high=0.2)
```

## Classification using percentile GSVA score thresholds
The package also provides a GSVA-based classification method 
`classify_GSVA_percent`, which uses percentile ranks as thresholds rather than 
absolute GSVA thresholds. The default percentile used by the function is 25%, so
the high expression abundance thresholds in both gene sets would be 75%, by 
default and low expression abundance thresholds in both gene sets would be 25%, 
default. A custom percentile threshold can be provided using the optional 
`percent_thresh` parameter. As for the absolute threshold function, a summary of 
the classification is printed to the console highlighting the number of samples 
in each pathway activity class ("Active", "Inactive" and "Uncertain") and the 
number of samples in total. Increasing the percentile threshold for 
classification has the equivalent effect of reducing the distance between the 
high expression abundance and low expression abundance thresholds for both gene 
sets, therefore reducing the frequency of "Uncertain" classifications.
```{r}
# default percentile = 25% (quartile)
classes_df.percent <- classify_GSVA_percent(ER_sig_df, data_se)
# custom percentile = 30%
classes_df.percent <- classify_GSVA_percent(ER_sig_df, data_se, 
                                            percent_thresh=30)
```

# Evaluating pathway activity classification
This package provides a method `calculate_accuracy` for evaluating pathway 
activity classification, which gets the actual labels and the labels classified
by the classification methods as parameters; then creates confusion matrix based
on them. The first parameter `true_labels_source` could be file-name, matrix or 
data frame and contains sample names and their corresponding actual 
labels.The second parameter `classes_df` could be matrix or data frame and 
contains classified labels found by the classification methods 
`classify_GSVA_percent` or `classify_GSVA_abs`. Third parameter `pathway` 
specifies with which pathway the labels(actual and classified) are  associated. 
Other parameters `display_statistics` and  `display_roc_curve` are boolean flags
used as optional features to display statistics and ROC Curve respectively.
Default values for the flags is FALSE.


```{r}
confusion_matrix <- calculate_accuracy("~/PathAnalyser/raw_data/Sample_labels.txt", 
                                       classes_df, "ER", display_statistics=TRUE, 
                                       display_roc_curve=TRUE)
```
# Visualising pathway activity
## Generating PCA plot
```{r}
classes_pca(normalized_se, classes_df, pathway_name = "ER")
```

# Example: Assessing ER pathway activity in a breast cancer RNAseq dataset
To demonstrate the standard workflow of assessing pathway activity in a 
transcriptomic data set, we will use PathAnalyser to classify samples in an RNA
seq data set according to ER pathway activity. The data set collected from The 
Genome Atlas Project (TCGA) contains raw read counts for over 20,000 genes for 
20 primary breast cancer samples.

## Read RNASeq gene expression dataset and ER pathway signature data
First the gene expression matrix data set text file which contains raw read 
counts and gene signature text files (up-regulated and down-regulated gene set 
files) are read into the expression matrix and gene expression data frame data 
types respectively.
```{r}
# Load transcriptomic data set (gene expression matrix of samples)
data_se <- read_expression_data("~/PathAnalyser/raw_data/toy_data.txt")
dim(data_se)
# read signature data from the two individual gene set files for up-regulated
# and down-regulated gene sets
ER_sig <- read_signature_data("~/PathAnalyser/raw_data/ESR1_UP.v1._UP.grp", "~/PathAnalyser/raw_data/ESR1_DN.v1_DN.grp")
```

## QC and pre-processing of the expression dataset and ER signature data
As the gene expression matrix contains raw RNASeq counts the data must be 
normalized prior to passing the matrix to any of the classification functions. 
Normalization of read counts is necessary to account for differences in 
sequencing depths amongst the libraries. The PathAnalyser `log_cpm` function can 
be used to perform a log counts per million (CPM) transformation on the gene 
expression matrix. Furthermore, `log_cpm` graphically confirms the logCPM 
transformation has been performed correctly since the boxplots 
displayed by the function, representing the logCPM values for each sample, are 
more aligned following the logCPM transformation compared to before the
transformation. 
```{r}
# Transforming matrix with log cpm transformation and sanity check of the transformation
normalized_se <- log_cpm_transformation(data_se)
```
To ensure that the gene names in the gene signatures and gene 
expression dataset are consistent, we use the `check_signature_vs_dataset` gene 
names to filter out genes from the gene expression matrix that are not in 
signature and those genes which are not expressed in at least 10% of samples.
```{r}
normalized_se <- check_signature_vs_dataset(normalized_se, ER_sig)
```


## Classifying breast cancer samples based on ER pathway activity.
Samples in a gene expression matrix can be classified according to evidence of 
expression consistency with the up-regulated and down-regulated gene sets of 
GSVA, by using absolute thresholds for high and low expression for each gene set 
of the signature via `classify_GSVA_abs` or percentile threshold for high and 
low expression for each gene set of the signature via `classify_GSVA_percent` 
(default=25%). The algorithm is described in detail.
```{r}
# classify samples using a percentile threshold (here 25% - quartile threshold
# is used as default
classes_df <- classify_GSVA_percent(ER_sig, normalized_se)
# using absolute thresholds
classes_df_abs <- classify_GSVA_abs(ER_sig, normalized_se, up_thresh.low = -0.2, 
                                    up_thresh.high = 0.2, dn_thresh.low = -0.2, 
                                    dn_thresh.high = 0.2)
```

## Visualizing ER pathway activity classification of the breast cancer samples
```{r}
classes_pca(normalized_se, classes_df, pathway_name = "ER")
```

## Assessing the accuracy of the pathway-based classification
```{r}
# read true pathway activity class labels for data set samples
true_labels_df <- read.table("~/PathAnalyser/raw_data/Sample_labels.txt", 
                             header=TRUE, sep = "\t")
# assess accuracy and generate confusion matrix for classification
confusion_matrix <- calculate_accuracy(true_labels_df, classes_df, 
                                       pathway = "ER", display_statistics = T)
```

# Session Info
The output of `sessionInfo` upon which this file was generated:
```{r}
sessionInfo()
```

# References
Hänzelmann, S., Castelo, R. & Guinney, J. GSVA: gene set variation analysis for 
microarray and RNA-Seq data. BMC Bioinformatics 14, 7 (2013). 
https://doi.org/10.1186/1471-2105-14-7

Symmans, W.F., Hatzis, C., Sotiriou, C., Andre, F., Peintinger, F., Regitnig, 
P., Daxenbichler, G., Desmedt, C., Domont, J., Marth, C. and Delaloge, S., 
2010. Genomic index of sensitivity to endocrine therapy for breast cancer. 
Journal of clinical oncology, 28(27), p.4111. 
https://dx.doi.org/10.1200%2FJCO.2010.28.4273
