---
output: html_document
editor_options: 
  chunk_output_type: console
---
Demo script for PathAnalyser R package which provides functionality for assessing ER and HER2 pathway activity
in breast cancer transcriptomic datasets.
 
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)
```

# Installation and setup
## Installing dependencies

```{r eval=F}
# If not already installed
install.packages("BiocManager")

BiocManager::install(c("GSVA", "pROC", "edgeR", "reshape2", "ggplot2","limma", 
                       "VennDiagram", "NCmisc", "futile.logger"), 
                     dependencies = TRUE)

```
## Installing package using devtools

```{r eval=F}
install.packages("devtools")
devtools::install_github("a-thind/PathAnalyser")
```

Alternatively, you can clone the GitHub repository in a Linux / MacOS terminal:
```{bash eval=F}
git clone git@github.com:a-thind/PathAnalyser.git
```

Then type the following in R to install a local source version of the package:

```{r eval=FALSE}
library(utils)
install.packages("~/PathAnalyser", repo=NULL, type="source")
```

Once installation is completed, load the library to start using PathAnalyser:
```{r setup}
library(PathAnalyser)
```


The `read_expression_data` function reads gene expression data from  
a file with an delimiter, the function checks the delimiter of the
file and reads the file accordingly. It returns an output with gene symbols/IDs as the row names and columns representing each sample 
IDs in the dataset.
```{r}
ER_expression_set <- read_expression_data("../raw_data/toy_data.txt")
# ER_expression_set<-ER_expression_set[,1:200]
# print(ER_expression_set[1:5, 1:5])
```

ER_expression_set contains RNASeq raw read counts for 20 primary 
breast cancer samples, 10 of which have ER pathway activity (ER+) and 10 which 
have inactive ER pathway activity (ER-).

```{r}
dim(ER_expression_set)
# Expression data for first 6 genes
head(ER_expression_set)
```

Gene signature files can be read by the `read_signature_data` function. It returns a dataframe with signature IDs/symbols in the
first column and their expression pattern is represented as up r
regulated or down regulated by +1 and -1 respectively.

```{r}
er_sign_df <- read_signature_data("../raw_data/ESR1_UP.v1._UP.grp", "../raw_data/ESR1_DN.v1_DN.grp")

dim(er_sign_df)
head(er_sign_df)
# up-regulated genes are given an expression value of 1
er_up_sign <- er_sign_df[er_sign_df$expression == 1 ,]
dim(er_up_sign)
# Down-regulated genes are given an expression value of -1
er_dn_sign <- er_sign_df[er_sign_df$expression == -1 ,]
dim(er_dn_sign)

```

## Quality control and pre-processing of data
### Log Counts Per Million (CPM) transformation of RNASeq expression matrices
The log_cpm_transformation function transforms the raw data by the method of log 
CPM and returns the transformed data. It also performs sanity check of the 
transformation by returning box plots for visualization of distribution of data 
before and after transformation.

```{r}
# using toy data set as expression matrix
data_se <- HER2_data_se1 
normalized_se <- log_cpm_transformation(ER_expression_set)
```

### Checking and filtering genes from the gene expression matrix
The gene names are first checked for inclusion in the gene signature data frame.
Genes that are included in the gene signature are retained then subjected to 
further filtration, in which their expression across the data set measured. If 
a gene is not present in at least 10% of the total number of samples in the 
data set, the gene is dropped from the expression matrix. A console message is
printed reporting the number of features (genes) retained in the final 
normalized expression matrix.
```{r}
normalized_se <- check_signature_vs_dataset(normalized_se, er_sign_df)
```

## Visualising GSVA Score Distribution
 `gsva_scores_dist` method  visualizes the GSVA score 
distribution for the abundance of expression of the up-regulated and 
down-regulated gene sets of the gene signature for each sample. 
```{r}
gsva_scores_dist(er_sign_df, normalized_se)

# Add thresholds on plot
library(ggplot2)
data_threshs <- data.frame(Geneset=c("Up", "Down"), vline=c(-0.2, 0.2))
plot + geom_vline(xintercept=data_threshs$vline, linetype=2)
```

Shrinking the distance between the high and low expression thresholds would 
result in fewer "Uncertain" pathway activity classifications 

```{r relaxed, fig.cap="GSVA scores distributions of samples with more relaxed thresholds for assessing expression consistency of the down-regulated gene set (left) and the up-regulated gene set (right) with the pathway gene expression signature. Low and high expression abundance thresholds for the down-regulated gene set (left) are set to -0.1 and 0.1 respectively, while low and high expression abundance thresholds for the up-regulated gene set (right) are also set to -0.1 and 0.1 respectively. Data is generated from running GSVA on logCPM-normalised `ER_data_se1` expression dataset with ER signature (`ER_sig_df`)."}
# more relaxed thresholds, fewer uncertain labels
data_threshs <- data.frame(Geneset=c("Up", "Down"), 
                           vline=c(-0.1, 0.1))
plot + geom_vline(xintercept=data_threshs$vline, linetype=2)
```

Conversely, expanding the distance between the thresholds would increase the 
frequency of "Uncertain" pathway activity classifications, as the number of samples
meeting the thresholds for consistency and inconsistency would be reduced.

```{r stringent, fig.cap="Density plots for GSVA scores distributions for samples with relatively more stringent thresholds for assessing consistency of the up-regulated gene set (left) and down-rgulated gene set (right) with the pathway signature. The low and high expression abundance thresholds are -0.4 and 0.4 for both up-regulated and down-regulated gene sets respectively. Data is generated from running GSVA on logCPM-normalised `ER_data_se1` expression dataset with ER signature (`ER_sig_df`)."}
# more stringent thresholds, greater uncertain labels
data_threshs <- data.frame(Geneset=c("Up", "Down"), vline=c(-0.4, 0.4))
plot + geom_vline(xintercept=data_threshs$vline, linetype=2)
```

## Classification using absolute GSVA score thresholds

```{r}
classes_df <- classify_GSVA_abs(er_sign_df, normalized_se, 
                                    up_thresh.low=-0.2, up_thresh.high=0.2, 
                                    dn_thresh.low=-0.2, dn_thresh.high=0.2)
```

## Classification using percentile GSVA score thresholds

```{r}
# default percentile = 25% (quartile)
classes_df.percent <- classify_GSVA_percent(er_sign_df, normalized_se)
# custom percentile = 30%
classes_df.percent <- classify_GSVA_percent(er_sign_df, normalized_se, 
                                            percent_thresh=30)
```

# Evaluating pathway activity classification


```{r}
confusion_matrix <- calculate_accuracy("../raw_data/Sample_labels.txt", 
                                       classes_df.percent, "ER", display_statistics=TRUE, 
                                       display_roc_curve=TRUE)
```

# Visualising pathway activity

## Generating PCA plot
```{r}
classes_pca(normalized_se, classes_df.percent, pathway_name = "ER")
```

# Session Info
The output of `sessionInfo` upon which this file was generated:
```{r}
sessionInfo()
```

